third_party/tonic/dart_weak_persistent_value.h

```diff
@@ -0,0 +1,48 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef LIB_TONIC_DART_WEAK_PERSISTENT_VALUE_H_
+#define LIB_TONIC_DART_WEAK_PERSISTENT_VALUE_H_
+
+#include <memory>
+
+#include "third_party/dart/runtime/include/dart_api.h"
+#include "tonic/common/macros.h"
+
+namespace tonic {
+class DartState;
+
+// DartWeakPersistentValue is a bookkeeping class to help pair calls to
+// Dart_NewWeakPersistentHandle with Dart_DeleteWeakPersistentHandle even in
+// the case if IsolateGroup shutdown. Consider using this class instead of
+// holding a Dart_PersistentHandle directly so that you don't leak the
+// Dart_WeakPersistentHandle.
+class DartWeakPersistentValue {
```

Comment:
I would really like to see unit tests of this API. I don't have a specific recommendation of tests to crib from, but there are many _unittests.cc files around to look into and you can ask on the flutter-engine discord for help if needed.

Reply:
Hey, I've looked into adding unit tests.

Tonic is a standalone library that only uses the Dart API, and it manages a Dart VM. Therefore, unit tests should start a Dart VM do some things with it and shut it down. However, because Tonic is standalone and does not import from the Flutter namespace we cannot reuse Flutters existing logic for starting a Dart VM.

Concretely, suppose a test would look like the following

```c++
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "../dart_state.h"
#include "gtest/gtest.h"
#include "third_party/dart/runtime/include/dart_api.h"

namespace tonic {

TEST(DartState, IsShuttingDown) {
  Dart_InitializeParams params = {};
  params.version = DART_INITIALIZE_PARAMS_CURRENT_VERSION;
  // TODO(dacoharkes): Set params correctly.
  char* init_error = Dart_Initialize(&params);
  free(init_error);
  Dart_Isolate isolate = Dart_CurrentIsolate();
  DartState* dart_state = DartState::From(isolate);
  EXPECT_FALSE(dart_state->IsShuttingDown());

  if (Dart_CurrentIsolate() != nullptr) {
    Dart_ExitIsolate();
  }
  char* result = Dart_Cleanup();
  free(result);

  // TODO(dacoharkes): Do this in the callback, same as in Flutter.
  dart_state->SetIsShuttingDown();

  EXPECT_TRUE(dart_state->IsShuttingDown());
}

}  // namespace tonic
```

where would we find a `vm_snapshot_data` and `vm_snapshot_instructions` for starting the DartVM with? 

Relevant Flutter logic which we can't reuse in a Tonic test: https://github.com/flutter/engine/blob/f7c7b41f0a0c72163d749321d070880fcac6402d/runtime/dart_vm.cc#L404-L425

Any suggestions for how to start a Dart VM in the tonic namespace?

(Side note on tests to crib from: There are remnants of unit tests in `third_party/tonic/filesystem/tests/` but those files have imports that don't exist. Origin of those files is the following commit:  https://github.com/flutter/engine/commit/276481c4d30fec1b9b12ec8c4f2b5ab28789fc45.)

Reply:
I've added unit tests outside the tonic namespace in the tonic folder.